### ISSUE TITLE
[Intel MKL]: Singularity build fails, but same works within docker.

### ISSUE THREAD

**Issue by karthikvadla** [apptainer/singularity#2536]
## Version of Singularity: 2.6.1-dist

Hi Team, 

I'm trying to create a singularity definition file similar to this [Dockerfile](https://github.com/tensorflow/tensorflow/blob/f92510eb61a469c0ea4652f8cd2582f71a821d80/tensorflow/tools/docker/Dockerfile.devel-mkl) file. My final singularity definition file (Singularity.devel-mkl) looks as below.

```
Bootstrap: debootstrap
OSVersion: bionic
MirrorURL: http://us.archive.ubuntu.com/ubuntu/

%help
    This Singularity definition contains Intel(R) Optimizations for
    TensorFlow.

    Use below command to build your singularity container.

    sudo singularity build <mysingularity_name>.sif Singularty.devel-mkl

%files


%post -c /bin/bash

    TF_BUILD_VERSION=r1.13
    PYTHON="python"
    PYTHON3_DEV=""
    WHL_DIR="/tmp/pip"
    PIP="pip"

    apt-get update && apt-get install -y software-properties-common
    add-apt-repository universe

    apt-get update && apt-get install -y --no-install-recommends \
    ${PYTHON} \
    ${PYTHON}-dev \
    ${PYTHON}-pip \
    ${PYTHON}-setuptools \
    ${PYTHON}-wheel \
    build-essential \
    curl \
    git \
    libcurl3-dev \
    libfreetype6-dev \
    libhdf5-serial-dev \
    libpng-dev \
    libssl-dev \
    libzmq3-dev \
    openjdk-8-jdk \
    openjdk-8-jre-headless \
    pkg-config \
    rsync \
    software-properties-common \
    unzip \
    zip \
    zlib1g-dev \
    && \
    apt-get clean && \
    rm -rf /var/lib/apt/lists/*

    ${PIP} --no-cache-dir install \
    Pillow \
    h5py \
    ipykernel \
    jupyter \
    keras_applications \
    keras_preprocessing \
    matplotlib \
    mock \
    numpy \
    pandas \
    scipy \
    sklearn \
    && \
    ${PYTHON} -m ipykernel.kernelspec

    # Set up Bazel.

    # Running bazel inside a `docker build` command causes trouble, cf:
    #   https://github.com/bazelbuild/bazel/issues/134
    # The easiest solution is to set up a bazelrc file forcing --batch.
    echo "startup --batch" >>/etc/bazel.bazelrc
    # Similarly, we need to workaround sandboxing issues:
    #   https://github.com/bazelbuild/bazel/issues/418
    echo "build --spawn_strategy=standalone --genrule_strategy=standalone" \
        >>/etc/bazel.bazelrc

    # Install the most recent bazel release.
    BAZEL_VERSION=0.20.0
    cd /
    mkdir /bazel && \
    cd /bazel && \
    curl -H "User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36" -fSsL -O https://github.com/bazelbuild/bazel/releases/download/$BAZEL_VERSION/bazel-$BAZEL_VERSION-installer-linux-x86_64.sh && \
    curl -H "User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/57.0.2987.133 Safari/537.36" -fSsL -o /bazel/LICENSE.txt https://raw.githubusercontent.com/bazelbuild/bazel/master/LICENSE && \
    chmod +x bazel-*.sh && \
    ./bazel-$BAZEL_VERSION-installer-linux-x86_64.sh && \
    cd / && \
    rm -f /bazel/bazel-$BAZEL_VERSION-installer-linux-x86_64.sh

    # Download and build TensorFlow.
    mkdir -p /tensorflow
    cd /tensorflow

    # Download and build TensorFlow.
    # Enable checking out both tags and branches
    export TAG_PREFIX="v" && \
        echo ${TF_BUILD_VERSION} | grep -q ^${TAG_PREFIX}; \
        if [ $? -eq 0 ]; then \
            git clone --depth=1 https://github.com/tensorflow/tensorflow.git . && \
            git fetch --tags && \
            git checkout ${TF_BUILD_VERSION}; \
       else \
            git clone --depth=1 --branch=${TF_BUILD_VERSION} https://github.com/tensorflow/tensorflow.git . ; \
        fi

    yes "" | ${PYTHON} configure.py
    cp .bazelrc /root/.bazelrc

    CI_BUILD_PYTHON=${PYTHON}

    # Set bazel build parameters in .bazelrc in parameterized_docker_build.sh
    # Use --copt=-march values to get optimized builds appropriate for the hardware
    #   platform of your choice.
    # For ivy-bridge or sandy-bridge
    # --copt=-march="avx" \
    # For haswell, broadwell, or skylake
    # --copt=-march="avx2" \
    cp .bazelrc /root/.mkl.bazelrc
    echo "import /root/.mkl.bazelrc" >>/root/.bazelrc

    tensorflow/tools/ci_build/builds/configured CPU \
    bazel build --config=mkl -c opt \
    tensorflow/tools/pip_package:build_pip_package && \
    bazel-bin/tensorflow/tools/pip_package/build_pip_package "${WHL_DIR}" && \
    ${PIP} --no-cache-dir install --upgrade "${WHL_DIR}"/tensorflow-*.whl && \
    rm -rf /root/.cache
    # Clean up Bazel cache when done.

    # TensorBoard
    # EXPOSE 6006
    # IPython
    # EXPOSE 8888

    cd /root

%environment
    # Set to 1 to see verbose MKL output
    MKLDNN_VERBOSE=0

    # KMP Settings
    KMP_AFFINITY=granularity=fine,compact,1,0
    KMP_BLOCKTIME=0
    KMP_SETTINGS=0

    export MKLDNN_VERBOSE KMP_AFFINITY KMP_BLOCKTIME KMP_SETTINGS

#%test
    echo "MKL shared libs: $(ldd $(pip show tensorflow | grep Location | cut -d ":" -f2)/tensorflow/libtensorflow_framework.so | grep libmklml)"
    python -c "import tensorflow as tf; print(tf.pywrap_tensorflow.IsMklEnabled())"

%labels
    Maintainer "Karthik Vadla <karthik.vadla@intel.com>"

```

### Expected behavior

Singularity build successful.

### Actual behavior

```
ERROR: /tensorflow/tensorflow/core/kernels/BUILD:3206:1: C++ compilation of rule '//tensorflow/core/kernels:reduction_ops' failed (Exit 1)
In file included from external/eigen_archive/unsupported/Eigen/CXX11/Tensor:124:0,
                 from ./third_party/eigen3/unsupported/Eigen/CXX11/Tensor:1,
                 from ./tensorflow/core/kernels/reduction_ops_common.h:27,
                 from tensorflow/core/kernels/reduction_ops_sum.cc:16:
external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorReduction.h: In static member function 'static void std::_Function_handler<void(_ArgTypes ...), _Functor>::_M_invoke(const std::_Any_data&, _ArgTypes&& ...) [with _Functor = Eigen::internal::TensorExecutor<Expression, Eigen::ThreadPoolDevice, Vectorizable, Tileable>::run(const Expression&, const Eigen::ThreadPoolDevice&) [with Expression = const Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<std::complex<float>, 0, 1, long int>, 16, Eigen::MakePointer>, const Eigen::TensorReductionOp<Eigen::internal::SumReducer<std::complex<float> >, const Eigen::IndexList<Eigen::type2index<0> >, const Eigen::TensorMap<Eigen::Tensor<const std::complex<float>, 1, 1, long int>, 16, Eigen::MakePointer>, Eigen::MakePointer> >; bool Vectorizable = true; bool Tileable = false]::<lambda(Eigen::internal::TensorExecutor<const Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<std::complex<float>, 0, 1, long int>, 16, Eigen::MakePointer>, const Eigen::TensorReductionOp<Eigen::internal::SumReducer<std::complex<float> >, const Eigen::IndexList<Eigen::type2index<0> >, const Eigen::TensorMap<Eigen::Tensor<const std::complex<float>, 1, 1, long int>, 16, Eigen::MakePointer>, Eigen::MakePointer> >, Eigen::ThreadPoolDevice, true, false>::StorageIndex, Eigen::internal::TensorExecutor<const Eigen::TensorAssignOp<Eigen::TensorMap<Eigen::Tensor<std::complex<float>, 0, 1, long int>, 16, Eigen::MakePointer>, const Eigen::TensorReductionOp<Eigen::internal::SumReducer<std::complex<float> >, const Eigen::IndexList<Eigen::type2index<0> >, const Eigen::TensorMap<Eigen::Tensor<const std::complex<float>, 1, 1, long int>, 16, Eigen::MakePointer>, Eigen::MakePointer> >, Eigen::ThreadPoolDevice, true, false>::StorageIndex)>; _ArgTypes = {long int, long int}]':
external/eigen_archive/unsupported/Eigen/CXX11/src/Tensor/TensorReduction.h:801:9: internal compiler error: in emit_move_insn, at expr.c:3698
         values[i] = internal::InnerMostDimReducer<Self, Op>::reduce(*this, firstIndex + i * num_values_to_reduce,
         ^~~~~~
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
Target //tensorflow/tools/pip_package:build_pip_package failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 74.923s, Critical Path: 51.44s
INFO: 2018 processes: 2018 local.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully
```

### Steps to reproduce behavior

Singularity definition file is based of `Bootstrap: debootstrap`. I have followed instructions from Singularity [docs](https://www.sylabs.io/guides/2.5/user-guide/appendix.html#build-debootstrap).

1. Singularity Version: 2.6.1-dist, installed using these [instructions](https://www.sylabs.io/guides/2.5/user-guide/installation.html#adding-the-mirror-and-installing)
2. OS Version used to test: Ubuntu 16.04
3. Create a file named `Singularity.devel-mkl` and paste above definition file content.
4. Command: `sudo singularity build --sandbox ubuntu/ Singularity.devel-mkl`
5. After couple of minutes you should see `FAILED: Build did NOT complete successfully`

*NOTE:* To cross check i did test `docker build` for above Dockerfile, build was successful.

**Comment by karthikvadla** [comment]
As this works with `singularity 3.0.1` . Closing the issue.

**Comment by mrodozov** [comment]
Hi there, 
as we have exactly the same failure we narrowed it as related to gcc 7.3.1. 
When we build with 8.2 or 7.4 it goes fine. Maybe in your new container you have different minor gcc7 version since when we tried to apply your recipe (run the container with singularity 3.0.3) it failed the same way. Have in mind regardless of the container we control out gcc version and it only worked when we moved to 7.4.0 revision 268351. 
Hope this would help anyone
